### PR TITLE
Add minimal validation of h5ad files to scp-ingest-pipeline (SCP-4498)

### PR DESCRIPTION
**Pull request by jlchang**
As a baby step towards accepting additional file types, this modification to ingest pipeline checks that files ending in .h5ad can be opened by scanpy. This action checks that the file is minimally valid.

To test via command line, activate the scp-ingest-pipeline repo virtualenv.
then from the scripts directory of the scp-ingest-pipeline repo, perform this setup:

```
source ../scripts/setup_mongo_dev.sh <path to your Github token file>
unset BARD_HOST_URL
```

from ingest directory, try and ingest the following test cases:

1. Good input file
`python ingest_pipeline.py --study-id addedfeed000000000000000 --study-file-id dec0dedfeed1111111111111 --study-id 5d276a50421aa9117c982845 --study-file-id 5dd5ae25421aa910a723a337 ingest_h5ad --ingest-h5ad --h5ad-file ../tests/data/h5ad/test.h5ad`

Look for the following output from scanpy indicating successful reading of the h5ad file:

> AnnData object with n_obs × n_vars = 238 × 187 backed at '../tests/data/h5ad/test.h5ad'
>     obs: 'submission_id', 'biosample_id', 'biosample_type', 'library_preparation_protocol', 'library_preparation_protocol__ontology_label', 'donor_id', 'species', 'species__ontology_label', 'mouse_strain', 'mouse_strain__ontology_label', 'organ', 'organ__ontology_label', 'brain_region', 'brain_region__ontology_label', 'sex', 'organism_age', 'organism_age__unit', 'organism_age__unit__ontology_label', 'disease', 'disease__ontology_label', 'seurat_clusters', 'cell_type', 'cell_type__ontology_label'

2. non-h5ad file (bad)
`python ingest_pipeline.py --study-id addedfeed000000000000000 --study-file-id dec0dedfeed1111111111111 --study-id 5d276a50421aa9117c982845 --study-file-id 5dd5ae25421aa910a723a337 ingest_h5ad --ingest-h5ad --h5ad-file ../tests/data/h5ad/bad.h5`

Look for the following output indicating the file's format was not detected

> Unsupported file format. Allowed file types are: application/x-hdf5

3. truncated h5ad file (bad)
`python ingest_pipeline.py --study-id addedfeed000000000000000 --study-file-id dec0dedfeed1111111111111 --study-id 5d276a50421aa9117c982845 --study-file-id 5dd5ae25421aa910a723a337 ingest_h5ad --ingest-h5ad --h5ad-file ../tests/data/h5ad/bad.h5ad`

Look for the following output indicating the file was could not be read by scanpy

> Unable to read file, "../tests/data/h5ad/bad.h5ad" using scanpy.

This fulfills SCP-4498.

### CODE REVIEW

File: ingest/ingest_files.py [hunk 1]
```diff
@@ -227,9 +237,12 @@ def open_file(self, file_path, open_as=None, start_point: int = 0, **kwargs):
                     open_file,
                 )
         else:
-            raise ValueError(
-                f"Unsupported file format. Allowed file types are: {' '.join(self.allowed_file_types)}"
+            msg = (
+                f"Unsupported file format. Allowed file types are: "
```
Comment (eweitz) [inline ingest/ingest_files.py:241]: Revisiting this, I think users would usually consider "file type" to be synonymous with "file extension" -- which is not the case here.  Many users won't know what MIME types are off hand, but including "MIME" in the error message at least gives them a hook they could search, something slightly more actionable.

```suggestion
                f"Unsupported file format. Allowed file MIME types are: "
```

I don't feel strongly about it, just something for consideration.

File: ingest/ingest_files.py [hunk 2]
```diff
@@ -298,6 +311,15 @@ def open_pandas(self, file_path, file_type, **kwargs):
         else:
             raise ValueError("File must be tab or comma delimited")
 
+    def open_h5ad(self, file_path, **kwargs):
+        """Opens file as AnnData object """
+        try:
+            return sc.read_h5ad(file_path, backed='r')
+        except OSError as e:
+            msg = f"Unable to read file, \"{file_path}\", using scanpy."
```
Comment (eweitz) [inline ingest/ingest_files.py:319]: Formatting errors as "{general message} {specific thing}" helps parsing, both visually and programmatically.  The particular wording suggested below is less relevant than ensuring the specific is completely after the general.

```suggestion
            msg = f"Scanpy cannot read file \"{file_path}\""
```

**Review by ehanna4 — APPROVED** [review]
provisional thumbs up based on code changes and earlier discussion. 

I haven't been able to set up my env locally to test this yet and am going to pair with Devon next week to get our M1s good to go for python/ingest pipeline development but don't want to hold up this PR with my pending functional review. If i can get it working on my loaner laptop i will update this comment accordingly

File: ingest/ingest_pipeline.py [hunk 3]
```diff
@@ -26,8 +26,8 @@
 # Ingest dense file
 python ingest_pipeline.py --study-id 5d276a50421aa9117c982845 --study-file-id 5dd5ae25421aa910a723a337 ingest_expression --taxon-name 'Homo sapiens' --taxon-common-name human --ncbi-taxid 9606 --matrix-file ../tests/data/dense_matrix_19_genes_1000_cells.txt --matrix-file-type dense
 
-# Ingest loom file
-python ingest_pipeline.py  --study-id 5d276a50421aa9117c982845 --study-file-id 5dd5ae25421aa910a723a337 ingest_expression --matrix-file ../tests/data/test_loom.loom  --matrix-file-type loom --taxon-name 'Homo Sapiens' --taxon-common-name human
+# Ingest h5ad file
+python ingest_pipeline.py  --study-id 5d276a50421aa9117c982845 --study-file-id 5dd5ae25421aa910a723a337 ingest_h5ad --h5ad-file ../tests/data/test.h5ad
```
Comment (eweitz) [inline ingest/ingest_pipeline.py:30]: Out-of-scope here, but maybe worthwhile as part of the larger "simplify Ingest Pipeline" initiative this quarter:

The use of both different subparsers (e.g. `ingest_expression`, `ingest_h5ad`) _and_ different argument names for "top level file" (e.g. `--matrix-file`, `--h5ad-file`) in each subparser seems unnecessarily complex.  Would be simpler to have a consistently-named argument for top-level file (e.g. `--file`), and encapsulate the business rules around subparser-specific file validation into the respective subparser argument handling?

File: ingest/ingest_pipeline.py [hunk 4]
```diff
@@ -102,6 +103,7 @@
     from .clusters import Clusters
     from .expression_files.dense_ingestor import DenseIngestor
     from .expression_files.mtx import MTXIngestor
+    from .h5ad import H5adIngestor
```
Comment (eweitz) [inline ingest/ingest_pipeline.py:106]: Out-of-scope here, but maybe worthwhile as part of the larger "simplify Ingest Pipeline" initiative this quarter:

This kind of redundant `import` block is an obvious maintainability issue in Ingest Pipeline.  A code comment above notes that it exists due to code in `single_cell_portal` repo, i.e. the public SCP CLI / `manage_study.py`.  But I suspect there's a better solution, e.g. either:

1. Simply declare the public SCP CLI unsupported (I recall analytics that showed essentially zero use, prompting indefinite postponement of related work) and delete the redundant imports, or
2. Implement a way to declare the imports once, while still satisfying that public SCP CLI use case (and testing it)

File: ingest/monitor.py [hunk 5]
```diff
@@ -47,6 +47,7 @@ def setup_logger(logger_name, log_file, level=logging.DEBUG, format="default"):
 def log_exception(dev_logger, user_logger, exception):
     user_logger.critical(str(exception))
     dev_logger.exception(exception)
+    print(str(exception))
```
Comment (eweitz) [inline ingest/monitor.py:50]: I assume this isn't intended, so:

```suggestion
```

File: tests/test_h5ad.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+            "../tests/data/h5ad/bad.h5",
+            "addedfeed000000000000000",
+            "dec0dedfeed0000000000000",
+        )
+        # passing obtain_data function to assertRaises using lambda
+        # otherwise bad_input.obtain_data() is evaluated and triggers
+        # an exception before assertRaises gets called
+        self.assertRaisesRegex(
+            ValueError,
+            "Unsupported file format. Allowed file types are: application/x-hdf5",
```
Comment (eweitz) [inline tests/test_h5ad.py:57]:
```suggestion
            "Unsupported file format. Allowed file MIME types are: application/x-hdf5",
```

File: ingest/cli_parser.py [hunk 7]
```diff
@@ -331,6 +331,20 @@ def create_parser():
...
+    parser_h5ad = subparsers.add_parser(
+        "ingest_h5ad", help="Indicates that h5ad file is being ingested"
+    )
+
+    parser_h5ad.add_argument(
+        "--ingest-h5ad",
+        required=True,
+        action="store_true",
+        help="Indicates that ingest of h5ad file should be invoked",
+    )
```
Comment (eweitz) [inline ingest/cli_parser.py:344]: This strikes me as unnecessary.  If it's necessary, could you refine `help` to briefly (perhaps further) indicate the purpose of this flag as distinct from the subparser argument (`ingest_h5ad`).  If it's unnecessary, could you remove it?

File: ingest/ingest_pipeline.py (hunk 3 above)
Comment (jlchang) [reply ingest/ingest_pipeline.py:30]: I agree that simplifying the number of argument names could simplify things but for some ingest use cases, there is more than one "top-level file". Subsampling requires both cluster and metadata files. DE expects cluster, metadata and matrix files. 

Also the current business rules around handling files is (mostly) consolidated in the IngestFile class. Simplification as suggested above would entail implementing a different overall design. I'm not adverse to it, but it's out of scope for this PR for sure.

File: ingest/monitor.py (hunk 5 above)
Comment (jlchang) [reply ingest/monitor.py:50]: It's a PITA to have to visit a log file to figure out what caused a validation to fail... Would it be acceptable to enclose it in an "if bypass_mongo_writes():" conditional? It seems odd to import config into the monitor module to do so though :(

File: ingest/cli_parser.py (hunk 7 above)
Comment (jlchang) [reply ingest/cli_parser.py:344]: Historically, it seemed necessary for the "run_ingest" function to work (Eno had indicated that there wasn't a good way to find out what subparser was invoked without passing a subparser-unique argument).  I might have found an argparse workaround to that difficulty, I'll do some testing and see.

File: ingest/monitor.py (hunk 5 above)
Comment (eweitz) [reply ingest/monitor.py:50]: I'm personally fine with this, it's just a bit unconventional for us to print production stuff to the console in JS and Python.

So either your suggested conditional, a brief code comment, or merely your explanation above seem sufficient to me.

File: ingest/ingest_files.py (hunk 2 above)
Comment (jlchang) [reply ingest/ingest_files.py]: `d52dd5b`

File: ingest/ingest_pipeline.py (hunk 4 above)
Comment (jlchang) [reply ingest/ingest_pipeline.py:106]: Will add this to the list of "simplify Ingest Pipeline" discussion topics

File: ingest/monitor.py (hunk 5 above)
Comment (jlchang) [reply ingest/monitor.py]: Removing for now (importing config to enable the conditional would require testing so I'll consider adding it back in my next batch of work)

File: tests/test_h5ad.py (hunk 6 above)
Comment (jlchang) [reply tests/test_h5ad.py]: `d52dd5b`

File: ingest/cli_parser.py (hunk 7 above)
Comment (jlchang) [reply ingest/cli_parser.py:344]: Will look into this simplification in my next batch of coding for Ingest Pipeline (We don't have good test coverage of the CLI interface within this repo so the testing required is more extensive than I have time for in this sprint).

**Pull request merged**